### PR TITLE
Fix: #5974 Cursor jump to the end of the document when a collaborator deletes focusing paragraph

### LINKED ISSUE #5974: Bug: Cursor jump to the end of the document when a collaborator deletes focusing paragraph
Lexical version: 0.14.5

## Steps To Reproduce

1. Launch the Playground in collaboration mode
2. Open two windows (A and B)
3. In window A, select a line
4. In window B, delete the line that is selected in window A
5. The focus in window A moves to the end of the document

## The current behavior
In the video, cursor in the right window jump to the end of the document when the paragraph is deleted from the left window.
This issue forces the cursor to the end of the editor during collaborative editing, which disrupts smooth editing experience.
https://github.com/facebook/lexical/assets/111737064/ad9ead44-d2ad-435c-aeb3-148dc1cc54cf

## The expected behavior
The cursor should not move to the end of the document when the collaborator deletes focusing paragraph.

### PR DESCRIPTION
**Pull request by matsuyama-k1**
fixed #5974 

https://github.com/facebook/lexical/assets/111737064/9b4bccee-23d8-447c-aacd-4da0cdbfe659

### CODE REVIEW

**Comment by ivailop7** [comment]
@fantactuka could you look into this one, would really like to land this one before the next release.

File: packages/lexical-yjs/src/Utils.ts [hunk 1]
```diff
@@ -541,3 +544,62 @@ export function removeFromParent(node: LexicalNode): void {
...
+) {
+  const currentParagraphKeys = $getRoot().getChildrenKeys();
+  const selectedNode = currentEditorState._nodeMap.get(anchorNodeKey);
+  if (!selectedNode) {
+    $getRoot().selectStart();
+    return;
+  }
+
+  // retrieve the parent paragraph node for the currently selected node
+  const selectedParagraphNode = $findParentParagraphNode(
```
Comment (ivailop7) [inline packages/lexical-yjs/src/Utils.ts:560]: I think you can reuse:

const selectedParagraphNode = $findMatchingParent(selectedNode, (node) => $isParagraphNode(node));

Validate that it's the behaviour you are after - get the closest paragraph parent to 'selectedNode'

**Comment by StyleT** [comment]
This is great! Can you pls add a test for this fix? To prevent regression

File: packages/lexical-yjs/src/Utils.ts [hunk 2]
```diff
@@ -541,3 +544,62 @@ export function removeFromParent(node: LexicalNode): void {
...
+
+  // If collaborator deleted paragraphs including first paragraph
+  // or RootNode accidently set to the selectedParagraphNode
+  if (!prevParagraphNodeKey) {
+    $getRoot().selectStart();
+    return;
+  }
+
+  const prevParagraphNode = $getNodeByKey(prevParagraphNodeKey);
+  if (prevParagraphNodeKey in currentParagraphKeys && prevParagraphNode) {
```
Comment (KatsiarynaDzibrova) [inline packages/lexical-yjs/src/Utils.ts:580]: As `currentParagraphKeys` is an array, it should probably be `includes()` not `in`
```suggestion
  if (currentParagraphKeys .includes(prevParagraphNodeKey) && prevParagraphNode) {
```

Comment (matsuyama-k1) [reply packages/lexical-yjs/src/Utils.ts:580]: thank you!

File: packages/lexical-yjs/src/Utils.ts (hunk 1 above)
Comment (matsuyama-k1) [reply packages/lexical-yjs/src/Utils.ts:560]: thank you!

**Comment by matsuyama-k1** [comment]
I added a test and fixed!

**Comment by matsuyama-k1** [comment]
The pull request that copy & paste my code was created and passed....😢
is it ok? I am sad....

mine
https://github.com/facebook/lexical/pull/5975/files#diff-3d291634235852090f5aa491928d6d43fa366302bf0222b0629adf3580e64ab2R549

yours
https://github.com/facebook/lexical/pull/6070/files#diff-3d291634235852090f5aa491928d6d43fa366302bf0222b0629adf3580e64ab2R552

**Comment by ivailop7** [comment]
A more generic fix was merged here: https://github.com/facebook/lexical/commit/b5fa87d0c6e92fb38f9cb484a7340ab41eff1186

**Pull request closed without merging**